### PR TITLE
[OP#168] - Improvement of Manure Management Module

### PR DESCRIPTION
**Pull request by ahmedjoubest**
Scientific functions fully implemented from the mockups presented by Giuseppe.

### CODE REVIEW

**Comment by ahmedjoubest** [comment]
@elaouniyassine 

[manure_emissions_simulation_input_herd_level_data.csv](https://github.com/user-attachments/files/25154054/manure_emissions_simulation_input_herd_level_data.csv)
[manure_emissions_example_calculation (1).xlsx](https://github.com/user-attachments/files/25154055/manure_emissions_example_calculation.1.xlsx)
[manure_emissions_simulation_input_cohort_level_data.csv](https://github.com/user-attachments/files/25154053/manure_emissions_simulation_input_cohort_level_data.csv)

#### Implement a run_* function

Minimum test coverage for the scientific functions

#### Implement a run_* function

**Arguments:**
- manure_management_system_fraction (at the cohort level) -- saved as csv examples
- manure_management_system_factors (at the herd level) -- saved as csv examples
- directemissions_manure_input_cohort_level_data -- saved as csv examples
- diet_ash -- defaulting to 0.08
- urinary_energy_fraction -- defaulting to 0.04

**The validation required at the run level:**
**No need** for the validation of fraction = 1, it's already taken care of in the scientific functions
However, several validation should be implemented at the run level:
- For each herd id, MMSs should be unique
- For each herd id, MMSs should be the same between the two tables:
  - manure_management_system_fraction
  - manure_management_system_factors
- each herd id should have the right 6 cohorts: 
  - manure_management_system_fraction 
  - directemissions_manure_input_cohort_level_data 
- All the other relevant validations


#### List of things to improve

- [x] The checks are currently failing
- [x] Improve the validation functions of the scientific functions: typically, the validation of `calc_total_n2o_emissions` function. Create new validation files for scientific functions (more factorization)
- [x] Linting, clarity, comments, etc... (of the run function)
- [x] In the validation for scientific functions: Protection against NA (for all values, nothing should be missing).

**Comment by ahmedjoubest** [comment]
Thank you @elaouniyassine  for this great work! The run function and validations are robustly working now.

I deleted the test validation file, because in my understanding it's just for reviewing purpose? Please feel free to get it back if you want to keep it, and if you judge that it's useful and won't be hard to maintain overtime.

**Comment by ahmedjoubest** [comment]
@tempiog @lydialanzoni @dommens 

The manure module is now ready for your review (run and scientific functions).

It's working as expected:
- Scientific functions working in an agnostic way to the MMS of category "other"
- The category "burn" and "pasture" of the MMS is treated in a particular way, and it's allowed to be missing
- Appropriate Validation is now implemented. You can test edge cases like:
  - duplicate MMSs
  - Missing Data
  - Fractions not totaling to 1
  - etc.
- The run function working with the exact starting point planned in the mock data provided by Giuseppe
 
Few other points:

- Please note that variable definitions still need a full review from your side (for both scientific and run functions).
- Alignment of variable names and input ranges is still pending (waiting for your greenlight for the sharepoint)

Please take some time to go through the code and test both normal and edge cases, then approve the PR if everything looks good to you.

If you approve the PR, please do not merge it yet, we will merge it later in the correct order, given the other pending PRs, to avoid conflicts.

We can discuss this during the meeting scheduled for Tuesday.

**Comment by dommens** [comment]
Great progress and nicely organized now. Wondering about "allow.cartesian = T in run_directemissions_manure.R, line 86. Should not be necessary unless there are duplicates in fraction of factors table and potentially suppresses the data.table warning/error on merge.

**Comment by ahmedjoubest** [comment]
> Great progress and nicely organized now. Wondering about "allow.cartesian = T in run_directemissions_manure.R, line 86. Should not be necessary unless there are duplicates in fraction of factors table and potentially suppresses the data.table warning/error on merge.

Thanks @dommens for spotting this.

In practice, `allow.cartesian = TRUE` only disables the data.table safeguard that stops joins when the result would exceed `nrow(x) + nrow(i)` (i.e. a potential large multiplicative join). I’m not sure there is a valid reason to use it here, because keys should indeed be unique and flagged.

In our case, both `fractions` and `factors` tables are already validated upfront and required to be unique on the join keys by our validation functions, so this merge should never rely on `allow.cartesian` anyway.

You’re right — its use here is unnecessary. It is not harmful given our current validations, but it is redundant and can be removed.

Thanks for noticing.

@elaouniyassine, could you please remove it here and, in the same commit, clean it up everywhere else in the codebase (both in the recently improved functions and in the unchanged ones)?

File: R/run_directemissions_manure.R [hunk 1]
```diff
@@ -1,576 +1,289 @@
...
-                   fifelse(Animal_short == "PGS", 0.06, 0.08)
-               ),
-             by = .I
+  # --- Step 3: Volatile solids (VS) -------------------------------------------
+  cohort_level_data[
+    ,
+    volatile_solids := calc_volatile_solids(
+      dry_matter_intake = dry_matter_intake,
+      diet_digestibility_fraction = diet_digestibility_fraction,
+      urinary_energy_fraction = urinary_energy_fraction,
```
Comment (lydialanzoni) [inline R/run_directemissions_manure.R:291]: Could we still keep this type of condition here (I don't know if it should be placed here or in the run function)?
This would allow to differenciate the arguments by species as an assumption when running the GLEAM pipeline/run_manure function.
I believe that this would required to add `species_short` as required input in the dt

```
urinary_energy_fraction =
                   fifelse(Animal_short == "PGS", 0.02, 0.04),
                 diet_ash =
                   fifelse(Animal_short == "PGS", 0.06, 0.08)
               ),
```

File: R/run_directemissions_manure.R [hunk 2]
```diff
@@ -1,576 +1,289 @@
...
+  mms_fraction <- data.table::copy(manure_management_system_fraction)
+  mms_factors <- data.table::copy(manure_management_system_factors)
+
+  # Join cohort fractions with herd-level MMS factors
+  mms_data <- merge(
+    mms_fraction,
+    mms_factors,
+    by = c("herd_id", "manure_management_system"),
+    all.x = TRUE,
+    allow.cartesian = TRUE
```
Comment (ahmedjoubest) [inline R/run_directemissions_manure.R]: @elaouniyassine 
I am adding this thread to the review to be resolved (last comment from Dominik): https://github.com/APPLITICS/gleam/pull/115#issuecomment-3870160607

File: R/run_directemissions_manure.R (hunk 1 above)
Comment (ahmedjoubest) [reply R/run_directemissions_manure.R:291]: Thanks @lydialanzoni for your review.
It's possible, but it's a change in the scien tific logic that wasn't included in the mock examples. In the Excel file, the animals were not part of the user input. **Do you internally agree on this scientific change?**

@elaouniyassine 
Here is a suggestion on how you can implement this feature (Lydia I let you confirm the structure).
Since it's a scientific change, it should be directly go to the scientific logic.
- [ ] 1 additional arguments to the `calc_volatile_solids` function: "animal" (long format, then you convert to short inside the script)
- [ ] `urinary_energy_fraction` should be transformed to two arguments:
  - `urinary_energy_fraction_pgs` for pigs
  - `urinary_energy_fraction_others` for others
  - Then you'll have to manage the conditonning inside this function
- [ ] You adjust the run function accordingly:
  - Add the animals (long)
  - Replace `urinary_energy_fraction` with the 2 new introduced ones: `urinary_energy_fraction_others` and `urinary_energy_fraction_pgs`

Comment (lydialanzoni) [reply R/run_directemissions_manure.R:291]: The initial idea to keep the condition in the run/call of the function because it **should not represent a change in the scientific function**, but rather an assumption that we do in GLEAM.

This means that if the user wants to run the code only using the function itself, will be able to specify whatever UE and Ash, even by species (or by cohort), by simply adding that as a column in the input table.

That said, I am not sure that splitting the variable into two would be a good solution......what do you think?

This was previously discussed with @tempiog and was implemented with https://github.com/APPLITICS/gleam/pull/110

Comment (lydialanzoni) [reply R/run_directemissions_manure.R:291]: Potential solution could be the following, in bold the chuncks of code that could be added to mantain the behaviour but still keep the flexibility

**Run function**: implement the `species_short`-based conditions in the run function. This will required having `species_short` as column in the `directemissions_manure_input_cohort_level_data`.
```
run_directemissions_manure <- function(
    directemissions_manure_input_cohort_level_data,
    manure_management_system_fraction,
    manure_management_system_factors,
    **urinary_energy_fraction =
      quote(fifelse(species_short == "PGS", 0.02, 0.04)),
    diet_ash =
      quote(fifelse(species_short == "PGS", 0.06, 0.08))**
) {
```

**Chunk of code to make it flexible to `urinary_energy_fraction` and `diet_ash` provided as inputs**:  the chunks of code in bold allow to overwrite the function argument for `urinary_energy_fraction` and `diet_ash` with optional values provided in the input table (which could even be cohort-level).

  
```
  # --- Step 1: Validate inputs ------------------------------------------------
  validate_directemissions_manure_inputs(
    directemissions_manure_input_cohort_level_data = directemissions_manure_input_cohort_level_data,
    manure_management_system_fraction = manure_management_system_fraction,
    manure_management_system_factors = manure_management_system_factors
  )

  # --- Step 2: Prepare inputs -------------------------------------------------
  cohort_level_data <- data.table::copy(directemissions_manure_input_cohort_level_data)
  mms_fraction <- data.table::copy(manure_management_system_fraction)
  mms_factors <- data.table::copy(manure_management_system_factors)
  
  
  **# Use column if present, otherwise evaluate default expression
  if (!"urinary_energy_fraction" %in% names(cohort_level_data)) {
    cohort_level_data[, urinary_energy_fraction := eval(urinary_energy_fraction)]
  }
  
  if (!"diet_ash" %in% names(cohort_level_data)) {
    cohort_level_data[, diet_ash := eval(diet_ash)]
  }**
  
```

**### Example**
In practice, after implementing the changes described above it would work like this:

###### **Usecase 1 . `urinary_energy_fraction` and `diet_ash` not provided in the input table.** 

- I would still need to have `species_short` as input of `directemissions_manure_input_cohort_level_data`

```
directemissions_manure_input_cohort_level_data <- data.table::fread(
  system.file(
    "extdata/examples/directemissions_manure_input_cohort_level_data.csv",
    package = "gleam"
  )
)

directemissions_manure_input_cohort_level_data[herd_id %in% c("1", "2", "3", "4"), species_short:="PGS"]
directemissions_manure_input_cohort_level_data[herd_id %in% c("5", "6"), species_short:="CTL"]

result <- run_directemissions_manure(
  directemissions_manure_input_cohort_level_data = directemissions_manure_input_cohort_level_data,
  manure_management_system_fraction = manure_management_system_fraction,
  manure_management_system_factors = manure_management_system_factors
)
```
###### **Usecase 2 . `urinary_energy_fraction` and `diet_ash provided in the input table as inputs from an advanced user** 

```
directemissions_manure_input_cohort_level_data[species_short=="PGS", urinary_energy_fraction:=0.08]
directemissions_manure_input_cohort_level_data[species_short=="CTL", urinary_energy_fraction:=0.02]
directemissions_manure_input_cohort_level_data[species_short=="PGS", diet_ash:=0.03]
directemissions_manure_input_cohort_level_data[species_short=="CTL", diet_ash:=0.03]

result_2 <- run_directemissions_manure(
  directemissions_manure_input_cohort_level_data = directemissions_manure_input_cohort_level_data,
  manure_management_system_fraction = manure_management_system_fraction,
  manure_management_system_factors = manure_management_system_factors
)
```

You can run the code and check the behaviour of the function with the two usecases:
```
result$volatile_solids
result_2$volatile_solids
```

**Comment by tempiog** [comment]
> @tempiog @lydialanzoni @dommens
> 
> The manure module is now ready for your review (run and scientific functions).
> 
> It's working as expected:
> 
> * Scientific functions working in an agnostic way to the MMS of category "other"
> * The category "burn" and "pasture" of the MMS is treated in a particular way, and it's allowed to be missing
> * Appropriate Validation is now implemented. You can test edge cases like:
>   
>   * duplicate MMSs
>   * Missing Data
>   * Fractions not totaling to 1
>   * etc.
> * The run function working with the exact starting point planned in the mock data provided by Giuseppe
> 
> Few other points:
> 
> * Please note that variable definitions still need a full review from your side (for both scientific and run functions).
> * Alignment of variable names and input ranges is still pending (waiting for your greenlight for the sharepoint)
> 
> Please take some time to go through the code and test both normal and edge cases, then approve the PR if everything looks good to you.
> 
> If you approve the PR, please do not merge it yet, we will merge it later in the correct order, given the other pending PRs, to avoid conflicts.
> 
> We can discuss this during the meeting scheduled for Tuesday.


Dear all, 
I reviewed and tested the new manure module and I have only few comments:

- [x] I agree with Lydia regarding the **diet_ash** and **urinary_energy_fraction**; these was part of a separate work package, and in my mockup I included only the example of one herd_id, but the idea for the time being is to not provide the defaults in the scientific functions (so that it could be used without assumptions on this), while providing them in the run function. If you agree, the solution suggested by Lydia could work. @ahmedjoubest and @elaouniyassine, do you see any issue with what Lydia proposed from the coding side? If you agree, I can implement the changes with a commit myself. 

- [x] @ahmedjoubest and @elaouniyassine, I tested using mms fractions that do not sum to 1; the code provides an error message, as expected, but I wonder it there is a way to report also which herd_id is responsible for the error, to help users correct it. 

- [x] parameters **ch4_max_producing_capacity_bo**, **n2o_ef3**, **n2o_ef4**, and **n2o_ef5** should have a maximum value allowed, but this will be defined in the sharepoint. @lydialanzoni could you review the sharepoint file for this? 

- [x] variable **ration_N2ON_to_N2O** should be called **ratio_N2ON_to_N2O**. I introduced the typo in my mockup, apologies for that. I can correct this with a commit myself. 

- [ ] Some corrections to the documentation will be required. I need to briefly discuss this with @lydialanzoni and will provide required corrections asap. OP User Story #287

**Comment by lydialanzoni** [comment]
@tempiog 

> * parameters **ch4_max_producing_capacity_bo**, **n2o_ef3**, **n2o_ef4**, and **n2o_ef5** should have a maximum value allowed, but this will be defined in the sharepoint. @lydialanzoni could you review the sharepoint file for this?

**SharePoint updated ✅** : https://unfao.sharepoint.com/:x:/s/GLEAM-X/IQC2BezjTpojRo8LXgthE65uAUuP8Cz3m1JDpn83vQYmMMs?e=ziqpf7

**Review by dommens — APPROVED** [review]

**Comment by ahmedjoubest** [comment]
@lydialanzoni @tempiog @dommens 

Thank you for your feedback. Giusappe, I'm using this thread to follow-up and not OP tasks to avoid duplicates.

Let's discuss the following points in 2 hours.

---

### Regarding Lydia's suggestion about diet_ash and urinary_energy_fraction

The only concern I have with Lydia’s approach is that a *scientific* decision is currently made at the `run_*()` level.
I understand that conditioning default values by species is itself an assumption (in GLEAM), but it still belongs to the *scientific layer* of the code, not the data-wrangling / orchestration layer. Isn't it a scientifi assumption?
That said, this approach is not wrong, it is mainly a design preference, to keep a clear separation between scientific logic and run-level logic. I’m fine proceeding this way if we agree on it.

If we go with this solution, a few important points to handle in the implementation:

* The default argument
  `urinary_energy_fraction = quote(fifelse(species_short == "PGS", 0.02, 0.04))`
  will not work, because `species_short` does not exist at function call time.
  You will need to:
  * add the species column to `directemissions_manure_input_cohort_level_data`, and
  * move the logic that derives `urinary_energy_fraction` and `diet_ash` from species inside the `run_*()` function, only when those columns are not explicitly provided.
  * If the columns are provided, they should be used directly (as in your current `#Use column if present, otherwise evaluate default expression` logic, with the species-based logic added there).

* In validation:

  * require a species column **only if** `urinary_energy_fraction` and `diet_ash` are not provided as columns,
  * but do not require species if both columns are already present.

* For species, please use the long format (as in the feed module).
  The idea discussed with Yassine is that package users only interact with the long format (more user-friendly).

* Please add a few basic tests for this behaviour. Then we can run a quick review.

---

> I tested using mms fractions that do not sum to 1; the code provides an error message, as expected, but I wonder it there is a way to report also which herd_id is responsible for the error, to help users correct it.

We can indeed add this logic. But it's a more global technical decision to make in the upcoming release: The current validation function doens't report to `herd_id` level at all. The most concrete one is the range inputs. If a value is out of range, it's blocked with a clear message but without reporting to the `herd_id` level. My recommendation given the timeline is to skip it for now, and add it as a high priority pre-release backlog item.
We can see with Dominik live today about this.

--- 

> parameters ch4_max_producing_capacity_bo, n2o_ef3, n2o_ef4, and n2o_ef5 should have a maximum value allowed, but this will be defined in the sharepoint. @lydialanzoni could you review the sharepoint file for this?

We had in the meantime the greenlight from Lydia for the sharepoint. This will be implemented before our meeting.

---

> variable ration_N2ON_to_N2O should be called ratio_N2ON_to_N2O. I introduced the typo in my mockup, apologies for that. I can correct this with a commit myself.

This has been corrected by Yassine in the very last commit before this comment: https://github.com/APPLITICS/gleam/pull/115/commits/8e6cff38a1653585cdb3dc2cfea2861b9ad32591

---

> Some corrections to the documentation will be required. I need to briefly discuss this with @lydialanzoni and will provide required corrections asap. OP User Story #287

Ok! I let you manage this internally and commit definitions updates directly in the code.

---

A small global comment about variable names: I really like the way you're currently trying to make everything as explicit as possible. Yassine have a small comment that we can discuss live.

**Comment by ahmedjoubest** [comment]
@tempiog @lydialanzoni @dommens 

All the discussed changes are now implemented:
- Implement manure validation ranges
- Turn diet ash/urinary fraction into primary args

This module is now completely ready for your:
- Review & Testing
- Full documentation of functions and variable definitions

We will then be able to merge it to staging after we merge the previous pending ones.

**Comment by tempiog** [comment]
> @tempiog @lydialanzoni @dommens
> 
> All the discussed changes are now implemented:
> 
> * Implement manure validation ranges
> * Turn diet ash/urinary fraction into primary args
> 
> This module is now completely ready for your:
> 
> * Review & Testing
> * Full documentation of functions and variable definitions
> 
> We will then be able to merge it to staging after we merge the previous pending ones.

Thank you @ahmedjoubest and @elaouniyassine,
I reviewed and tested the changes, and everything works as discussed. 
@lydialanzoni we can work on the documentation refinement and commit changes in this branch.

**Review by lydialanzoni — APPROVED** [review]

**Review by tempiog — APPROVED** [review]
The documentation was updated and reviewed. From our side, the branch is ready for the merge.
